File: python/verifying_an_alien_dictionary.py
class Solution:
    def isAlienSorted(self, words: list[str], order: str) -> bool:
        """time complexity: O(n * m), where n is the number of words and m is the length of the longest word."""
        order_id: dict[str, int] = {c: i for i, c in enumerate(order)}
        for i in range(len(words) - 1):
            w1, w2 = words[i], words[i + 1]
            for j in range(len(w1)):
                if j == len(w2):
                    return False
                if w1[j] != w2[j]:
                    if order_id[w1[j]] > order_id[w2[j]]:
                        return False
                    break
        return True


def test_isAlienSorted_case_1():
    # arrange
    words: list[str] = ["hello", "leetcode"]
    order: str = "hlabcdefgijkmnopqrstuvwxyz"
    expected: bool = True

    # act
    solution = Solution()
    actual: bool = solution.isAlienSorted(words, order)

    # assert
    assert expected == actual


def test_isAlienSorted_case_2():
    # arrange
    words: list[str] = ["word", "world", "row"]
    order: str = "worldabcefghijkmnpqstuvxyz"
    expected: bool = False

    # act
    solution = Solution()
    actual: bool = solution.isAlienSorted(words, order)

    # assert
    assert expected == actual


def test_isAlienSorted_case_3():
    # arrange
    words: list[str] = ["apple", "app"]
    order: str = "abcdefghijklmnopqrstuvwxyz"
    expected: bool = False

    # act
    solution = Solution()
    actual: bool = solution.isAlienSorted(words, order)

    # assert
    assert expected == actual

File: rust/src/code/verifying_an_alien_dictionary.rs
struct Solution;

impl Solution {
    pub fn is_alien_sorted(words: Vec<String>, order: String) -> bool {
        use std::collections::HashMap;

        let order: Vec<char> = order.chars().collect();
        let mut order_idx: HashMap<char, usize> = HashMap::new();
        for i in 0..order.len() {
            order_idx.insert(order[i], i);
        }
        for i in 0..words.len() - 1 {
            let w1: Vec<char> = words[i].clone().chars().collect();
            let w2: Vec<char> = words[i + 1].clone().chars().collect();
            for j in 0..w1.len() {
                if j == w2.len() {
                    return false;
                }
                if w1[j] != w2[j] {
                    if order_idx[&w1[j]] > order_idx[&w2[j]] {
                        return false;
                    }
                    break;
                }
            }
        }
        true
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test_is_alien_sorted_case_1() {
        // arrange
        let words: Vec<String> = vec!["hello".to_string(), "leetcode".to_string()];
        let order: String = "hlabcdefgijkmnopqrstuvwxyz".to_string();
        let expected: bool = true;

        // act
        let actual = Solution::is_alien_sorted(words, order);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_is_alien_sorted_case_2() {
        // arrange
        let words: Vec<String> = vec!["word".to_string(), "world".to_string(), "row".to_string()];
        let order: String = "worldabcefghijkmnpqstuvxyz".to_string();
        let expected: bool = false;

        // act
        let actual = Solution::is_alien_sorted(words, order);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_is_alien_sorted_case_3() {
        // arrange
        let words: Vec<String> = vec!["apple".to_string(), "app".to_string()];
        let order: String = "abcdefghijklmnopqrstuvwxyz".to_string();
        let expected: bool = false;

        // act
        let actual = Solution::is_alien_sorted(words, order);

        // assert
        assert_eq!(expected, actual);
    }
}
